### PR TITLE
ICU-21021 Changing MeasureUnit::splitToSingleUnits to return std::pair

### PR DESCRIPTION
**Pull request by sffc**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21021
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

**Review by markusicu — DISMISSED** [review]
lgtm

**Comment by markusicu** [comment]
Oh wait, lots of build bots are unhappy. Please check.

File: icu4c/source/i18n/number_usageprefs.cpp [hunk 1]
```diff
@@ -94,13 +94,12 @@ void mixedMeasuresToMicros(const MaybeStackVector<Measure> &measures, DecimalQua
         U_ASSERT(micros->outputUnit.getComplexity(status) == UMEASURE_UNIT_MIXED);
         U_ASSERT(U_SUCCESS(status));
         // Check that we received measurements with the expected MeasureUnits:
-        int32_t singleUnitsCount;
-        LocalArray<MeasureUnit> singleUnits =
-            micros->outputUnit.splitToSingleUnits(singleUnitsCount, status);
+        MeasureUnitImpl temp;
+        const MeasureUnitImpl& impl = MeasureUnitImpl::forMeasureUnit(micros->outputUnit, temp, status);
```
Comment (hugovdm) [inline icu4c/source/i18n/number_usageprefs.cpp:98]: Hmm, what's the best resolution for this:

```
number_usageprefs.cpp:98:32: error: unused variable 'impl' [-Werror,-Wunused-variable]
        const MeasureUnitImpl& impl = MeasureUnitImpl::forMeasureUnit(micros->outputUnit, temp, status);
```

This impl is only used in U_ASSERT, which is why I placed the whole block in a U_DEBUG. However I now see U_DEBUG can be defined while U_ASSERT is skipped:

```
#if U_DEBUG
#   include <assert.h>
#   define U_ASSERT(exp) assert(exp)
#elif U_CPLUSPLUS_VERSION
#   define U_ASSERT(exp) (void)0
#else
#   define U_ASSERT(exp)
#endif
```

I must have missed a warning for `singleUnits` then?

Can one have multiple statements together in a U_ASSERT? Perhaps drop the U_ASSERT and rely only on the U_DEBUG guards?

**Comment by hugovdm** [comment]
```
            unitMeasure {
            FAIL: :0: Mixed unit: Unsafe Path (U_INTERNAL_PROGRAM_ERROR)
            FAIL: Mixed unit: Unsafe Path; got ""; expected "3 yd, 1 ft, 11.4 in"
            FAIL: :0: Mixed unit: Safe Path (U_INTERNAL_PROGRAM_ERROR)
            FAIL: Mixed unit: Safe Path; got ""; expected "3 yd, 1 ft, 11.4 in"
            assertFormatSingle destructor: expected success but got error: U_INTERNAL_PROGRAM_ERROR scope: Mixed unit
```

(I can't quickly imagine off the top of my head what the error might be.)

**Comment by sffc** [comment]
I think I must be letting some memory get corrupted somewhere with my new code in the units router. I'll take a closer look in Valgrind this afternoon.

**Comment by sffc** [comment]
OK, let's see if the above amendment fixes the issue.

**Review by markusicu — APPROVED** [review]
lgtm

**Pull request merged**